### ISSUE TITLE
Overhaul of scoring concept

### ISSUE THREAD

**Issue by daf** [ioos/compliance-checker#37]
This issue meant for discussion, including but not limited to @dpsnowden, @rsignell-usgs, @kwilcox, @DanielJMaher, @lukecampbell, @oychang ...

The scoring system currently in the compliance-checker feels not up to par.  This issue attempts to explain how it got to be the way it is, and where to go from here.
### Initial Design

The initial design goals were this:
- Score a dataset by multiplying the weight (low=1, medium=2, high=3) of each top-level Result by the value (0 for 0, 0.5 for any fractional, 1 for 1)
- Provide a stable upper-bound per checker, not varying per dataset (aka same upper bound no matter how many variables)

The per variable checks rely on the concept of grouping, which results in the tree-like output you see today (and needs to be clarified in documentation).  Therefore, instead of something like the following:

```
var_temperature_units_valid passed
var_salinity_units_valid failed
```

You get something like this:

```
var
    temperature
        units
            valid  Passed
    salinity
        units
            valid  Failed
```

... and everything under the "var" tree counts as ONE top level Result to be added towards the total score.  In this case, "var" would be 1/2, or partial, multiplied by whatever is the maximum weight observed in this tree.  THIS IS COMPLICATED.
#### Grouped check contribution

A major downside to this is that grouped items contribute very little to the overall score.  Consider the ACDD checker: most of the checks are for global attributes, then about four checks per variable, which all get grouped together.

Since the ACDD checker was first completed, this heavily guided the scoring ideas.  As soon as the CF checker was taking shape, problems emerged: we do some number of global checks, then do a TON of per-variable checks, which barely contribute to the score. Simple global attribute checks should not carry more weight than the per-variable checks!
#### Optional checks (DSG)

[Section 9 of the CF document](http://cfconventions.org/Data/cf-convetions/cf-conventions-1.6/build/cf-conventions.html#discrete-sampling-geometries) defines DSGs.  Clearly, not every dataset going through the CF checker will or should have these properties or even perform these tests (we're fixing #36 to that affect).

Therefore, there is no way to give a consistent "CF checker" upper-bound between different datasets if one is DSG and the other is not.  It would have to be something like a "bolt-on" style score - "Dataset X scored 120/140 of normal CF, and 20/30 of DSG", almost treating it as if it was a second checker.
#### Split out CF/DSG checker

As mentioned in the optional section above, "optional" style checks can be implemented as a second checker.

Pros: 
- Consistent reporting
- Could solely report DSG compliance if wanted

Cons:
- Would this mean users would have to `--test=cf,cf-dsg` every time?
#### Do away with upper bound/scoring altogether

Instead, consider a "demerit" style scoring approach.  Instead of a "100/120, here's all your wrong things", just output the wrong things. For example:

```
CF: -6

HIGH: No cf_role detected (-3)
MEDIUM: No author attribute (-2)
LOW: Variable "temperature" is missing a comment about how important it is (-1)
```

Pros:
- Could possibly do away with grouping mechanism
- Still variable independent

Cons:
- Extremely negative style approach (HERE'S HOW YOU'RE WRONG, DUMMY)
- Provides no baseline other than Perfect (0)

<hr />

I apologize for the length, but needed to get this all recorded.  Please take a bit to digest and give some thoughts, or ask for clarification, as this is a not very well organized brain dump.

**Comment by mhidas** [comment]
Have there been any further thoughts on this topic since last year?

For the way we'd like to use of the checker within IMOS, we don't really need any kind of scoring system. All we want to know is what's wrong with the file, so we can send feedback to its creator and ask them to fix it. For a file that passes all checks, the only output we need is an exit status of 0 to determine what happens next. So for us, the "demerit" approach mentioned above would be perfect. If this doesn't work for other applications, it could at least be an optional output mode.

Also, I disagree with "Provides no baseline other than Perfect". There's already the `--criteria` switch to select the level/weight (low/medium/high) of checks that needs to be passed to make the final exit status 0 (i.e. Pass). So we really have 3 baselines, which can be thought of as 
* OK, all mandatory checks passed (`--criteria=lenient`)
* GOOD, all mandatory & recommended checks passed (`--criteria=normal`)
* PERFECT, all checks passed (`--criteria=strict`)

**Comment by abirger** [comment]
@mhidas , so, would the 3-baseline test with `--criteria=normal` result be FAILED or OK if just mandatory checks pass?

**Comment by mhidas** [comment]
@abirger In that case the overall result would be FAILED.

I'm thinking of the weight of each check result to mean 
* HIGH = mandatory check
* MEDIUM = recommended check
* LOW = "would be nice" check?

So if the highest weight of any failed check is shown in the top row, the overall result according to each criteria is

|         | HIGH | MEDIUM | LOW  |
|---------|------|--------|------|
| lenient | FAIL | OK     | OK   |
| normal  | FAIL | FAIL   | OK   |
| strict  | FAIL | FAIL   | FAIL |

**Comment by kwilcox** [comment]
I'm going on record in favor of a "demerit" style approach. Shaming people is already something I'm pretty good at. I haven't come across a use-case to knowing what is correct with a dataset... everyone always wants to know what is wrong with it.

**Comment by mhidas** [comment]
@kwilcox I'm with you on that. And if it is generally agreed that this is the way to go, perhaps the `Scoring Breakdown` section of the text output could be removed (unless verbose output is requested)?

**Comment by Bobfrat** [comment]
Hey guys, I'd like to resurrect this discussion and begin work on overhauling the scoring system in the next Milestone. It seems that in general folks are leaning towards the demerit style approach. Does anyone have any other suggestions or concerns about this approach? Or any ideas on how they would like the results presented?

@daf @lukecampbell @mhidas @kwilcox @benjwadams @brianmckenna @mwengren 
@jbosch-noaa @dpsnowden @abirger @martinjuckes

**Comment by daf** [comment]
Thanks for revisiting.  Thinking about this on and off for the last few years, plus seeing it used in the wild, I think @mhidas really hits the crux of the issue above:

> All we want to know is what's wrong with the file, so we can send feedback to its creator and ask them to fix it. 

I fully agree.  Trying to come up with a "score" is too difficult - checks that only run when certain conditions exist or checks that are conditional eg "does this attribute exist, is it in this set of values?" make this concept very hard.  The grouping mechanism was an attempt to normalize all of these issues but in implementation grossly overcomplicates writing checks.

I fully support the concept of the "demerit" or "here's your issues" style approach.  Keep the weight/priority concept - this will help categorize output at the end, as well as set failure levels (`--criteria`) as noted by @mhidas above.

Something like:

```
You have 4 REQUIRED issues:
- variable 'something' is missing attribute 'standard_name'
- ...

You have 2 RECOMMENDED issues:
- ...
```

This strips away a lot of complexity and gets down to the real goal of running this tool: "what do I need to do to fix my files?"

**Comment by kwilcox** [comment]
I'll chime in as a heavy user of the CC.

I agree with everything @daf says above. Keep in mind that someone won't always want to know exactly _what_ is wrong with a dataset but just that _something_ is wrong. I would not be opposed to another run option to "fail quick" in cases that a check has failed at or below the failure threshold of `--criteria`.

The return format of the checker should also be re-thought. Below is what I currently do to extract all of the relevant information.

```python
def check_dataset(args):
    check_suite = CheckSuite()
    check_suite.load_all_available_checkers()

    outhandle, outfile = tempfile.mkstemp()

    try:
        return_value, errors = ComplianceChecker.run_checker(
            ds_loc=args.file,
            checker_names=['gliderdac'],
            verbose=True,
            criteria='normal',
            output_format='json',
            output_filename=outfile
        )
        assert errors is False
        return 0
    except AssertionError:
        with open(outfile, 'rt') as f:
            ers = json.loads(f.read())
            for k, v in ers.items():
                if isinstance(v, list):
                    for x in v:
                        if 'msgs' in x and x['msgs']:
                            L.warning('{} - {}'.format(args.file, x['msgs']))
        return 1
    except BaseException as e:
        L.warning('{} - {}'.format(args.file, e))
        return 1
    finally:
        os.close(outhandle)
        if os.path.isfile(outfile):
            os.remove(outfile)
```

**Comment by mhidas** [comment]
:+1: Excellent, I'm glad everyone seems to be thinking the same way.

@kwilcox I had a play with your `check_dataset` function. I agree that it's more or less what we need. However I think the's a misunderstanding behind your exception handling. `ComplianceChecker.run_checker` returns `errors == True` only if *exceptions* occur during checking. It is NOT an indication of compliance issues with the file!
As it is you need to look a couple of levels down into the json structure before you can determine whether any checks failed.

**Comment by Bobfrat** [comment]
One thing I've found that frustrates users is when they fix an issue in their file and rerun the checker only to find that the fix just created another failed check. 

An example would be any of the date attributes in ACDD. If they omit the attribute they get docked for the fact its missing. So then they add the attribute to the file but now it is checked against the ISO format. It's best if they understand all the checks that will be run against that attribute on the first fail.

**Comment by Bobfrat** [comment]
Some more thoughts:

Let's make sure we don't break backwards compatibility. This will allow us to roll out a solution for the compliance checker first, then work on updating all the plugins.

Here's 2 other compliance checker web versions. I recommend taking a few minutes to check them out if you haven't already. I think there are elements we can take from each.
 
https://podaac-uat.jpl.nasa.gov/mcc/
http://pumatest.nerc.ac.uk/cgi-bin/cf-checker-dev.pl

Do you guys agree that the web version will likely get more users than the CLI?  Our web application will need an overhaul as well.

**Comment by mhidas** [comment]
> It's best if they understand all the checks that will be run against that attribute on the first fail.

True, but this can be documented in a static way elsewhere. It doesn't need to be repeated in the output every time a file is checked. The simplest thing to do is add a link to a site/document describing the relevant convention (#154 and https://github.com/ioos/compliance-checker/issues/186#issuecomment-153511916). As you suggest, details of the actual checks applied by the checker, would be even more helpful, but again, this is fixed for each check suite, not dependent on the file being checked (except of course for conditional checks like DSG...)

The http://pumatest.nerc.ac.uk/cgi-bin/cf-checker-dev.pl reports its results closer to the way I would, i.e. focusing on things that can be fixed and not reporting things that are already compliant.

Within IMOS / AODN we use the checker as part of our automated data ingestion process. It is called as a function from within Python code. We also use the CLI. We are currently not using the web app, though we may set up our own version of it, with our plugins, at some point in the future.

**Comment by Bobfrat** [comment]
@daf @lukecampbell @mhidas @kwilcox @benjwadams @brianmckenna @mwengren
@jbosch-noaa  @abirger @ocefpaf 

Hey guys, we have been working on the scoring overhaul and would like your feedback. All of the code for the new scoring output is on the ```refactor-scoring``` branch. Please pull it down and take it for a test drive. Here are some of the new features:

- Removal of the scoring section for each output format. Left with only the corrective actions based on the criteria level passed in
- High, Medium, Low Priority headers have been replaced with the terminology for each check (i.e. Highly Recommended, Recommended, and Suggested for ACDD)
- Added ability to concatenate results from many input files into a single output file
- Added ability to export multiple formats (html, text, json) in one run
- Removed concept of score grouping which had unexpected scoring consequences

There are examples of the new functionality on the [README](https://github.com/ioos/compliance-checker/blob/refactor-scoring/README.md) on that branch

Once we're happy, we'll make a release and update the web version to show the same output. I'll be sending an email with this info to the IOOS tech email list so I apologize for cross posting.

**Comment by Bobfrat** [comment]
The refactor-scoring branch was merged into master. I'll be creating a new release tomorrow unless anyone has any issues they want to raise.

**Comment by kwilcox** [comment]
Is there any documentation of the changes? Can we get a full version bump to `4.0.0` so I have some time to port over existing processes and stay on `3.x` for awhile?

**Comment by Bobfrat** [comment]
Yeah I'll make sure we bump to 4.0.0. Summary below (README has some examples and info as well):
- Removal of the scoring section for each output format. Left with only the corrective actions based on the criteria level passed in
- High, Medium, Low Priority headers have been replaced with the terminology for each check (i.e. Highly Recommended, Recommended, and Suggested for ACDD)
- Added ability to concatenate results from many input files into a single output file
- Added ability to export multiple formats (html, text, json) in one run
- Removed concept of score grouping which had unexpected scoring consequences

**Comment by jbosch-noaa** [comment]
Great! I plan on announcing this in the next Eyes on the Ocean newsletter.  @Bobfrat - Is the bulleted summary above a complete list of version updates to include in the announcement?

**Comment by Bobfrat** [comment]
Just released 4.0.0 with the scoring (or lack thereof) updates. Release notes [here](https://github.com/ioos/compliance-checker/releases/tag/4.0.0)

https://pypi.python.org/pypi/compliance-checker/4.0.0
